[team-10][Dumba, Min] MVVM 구조설계와 첫 번째 화면(Main) 구성

youngminshim-de opened the pull request:
### 작업내용
- MVVM 패턴을 적용하여 클린 아키텍쳐 구조를 지키려 노력하며 설계했습니다.
- UICollectionViewDiffableDataSource를 이용하여 CollectionView를 구성했습니다.
- Combine을 이용하여 View - ViewModel 간의 Data를 Binding 했습니다.
- Alamofire를 이용하여 Network 통신을 구현했습니다.
- Toast를 이용하여 Title을 터치하면 상품의 개수를 띄우도록 구현했습니다.

### 고민거리
- Title을 누르면 Toast를 띄워야하는 요구사항이 있어, BanchanCustomCellHeader(UICollectionReusableView) 에 
TapGestureRecognizer 를 추가해줌으로써, Action을 연결해 주었는데, 여기서 Cell의 개수가 필요하기에 cellCount라는 property를 선언해줌으로써 해결해주었습니다.

- 네트워크 통신 주체인 NetworkService Class가 현재 MVVM 패턴에서 Repository  미구현으로  Network 접근이 viewModel, view에서 등 여러곳에서 접근되어 지향하고 싶지만 어쩔 수 없이 싱글톤으로 구현하였는데 Repository의 윤곽이 잡히게 되면 싱글톤 방식에서 벗어날 수 있는지 고민하고 있습니다.

kihyuk-sung committed:
docs: Add Branch strategy into README.md

- README.md에 팀 브랜치 전략 추가.

kihyuk-sung committed:
docs: Add commit convention

- 커밋 규약 추가

kihyuk-sung committed:
docs: Add team members

- 팀명 추가
- 팀원 추가

youngminshim-de committed:
feat: Create Project

youngminshim-de committed:
feat: NavigationController & ViewController added

반찬 컬렉션뷰를 나타낼 BanchanListViewController 와 상세페이지를 나타내는 DetailBanchanViewController 추가

youngminshim-de committed:
chore: Folder Grouping

MVVM 패턴적용을 위한 폴더 그룹핑 & 주석제거

youngminshim-de committed:
feat: Add Banchan Model

반찬 모델 생성

youngminshim-de committed:
feat: Add DetailBanchan Model

상세반찬모델 생성

youngminshim-de committed:
feat: Add BanchanCustomCell

BanchanCustomCell.xib 파일에 Cell Layout 작성

issue: #12

youngminshim-de committed:
Add BanchanCustomCell Header

반찬컬렉션 뷰의 헤더파일 추가

ghis22130 committed:
feat: Add BanchanListViewModel, UICollectionViewDiffableDataSource

섹션별로 업데이트를 하기위해 UICollectionViewDiffableDataSource를 도입하였습니다.
이에 따른 NSDiffableDataSourceSnapshot을 적용해 주었고 가독성을 높이기 위해 typealias를 통하여 재정의 해주었습니다.
DataSource에 정보를 주기 위한 BanchanListViewModel을 정의해주었습니다.

issue: #13

youngminshim-de committed:
chore: move BanchanCustomCell class

반찬커스텀셀, 반찬커스템셀 헤더 다른 폴더로 이동

youngminshim-de committed:
chore: merge

youngminshim-de committed:
fix: Mod Cell Size

Cell Autolayout 수정

issue: #12

youngminshim-de committed:
Merge pull request #17 from ghis22130/feature-UI

Feature UI

youngminshim-de committed:
refactor: BanchanListViewModel refactoring

Section 클래스를 만들어 기존 BanchanListViewModel 의 역할을 대신함

youngminshim-de committed:
chore: Add Cocoapod & install Toaster

Toaster 패키지 cocoapod으로 설치

youngminshim-de committed:
Merge pull request #20 from ghis22130/feature-UI

BanchanListViewModel Refactoring & Install Toaster

ghis22130 committed:
chore: Install Alamofire CocoaPod

네트워크 프레임워크인 Almofire을 사용하기위해 Cocoapod을 통하여 설치하였습니다

issue: #16

ghis22130 committed:
feat: Add NetworkService Class

HTTPMethod를 통한 네트워크 통신을 Alamofire를 사용하여 작동하는 NetworkService Class의 초안을 구현하였습니다.

issue: #16

ghis22130 committed:
feat: Add BanchanListDTO, BanchanListItemDTO

네트워크와 통신할 DTO 객체를 구현하였습니다.
DTO 자체에서 toDomain() 메소드를 두어 Entity를 바로 생성해줄 수 있도록 하였습니다.

issue: # 16

ghis22130 committed:
feat: Add BanchanListDTO, BanchanListItemDTO

네트워크와 통신할 DTO 객체를 구현하였습니다.
DTO 자체에서 toDomain() 메소드를 두어 Entity를 바로 생성해줄 수 있도록 하였습니다.

issue: # 16

ghis22130 committed:
feat: Add FetchBanchanListUseCase

네트워크와 통신하여 반찬리스트를 받아올 UseCase를 작성하였습니다

issue: #16

ghis22130 committed:
feat: Add FetchImageUseCase

String URL로 부터 Image를 불러오기 위한 UseCase를 작성하였습니다.
UIKit을 모르게 하기 위해 가공된 Data에 대한 completion을 제공하여 Image생성하게 끔 하였습니다.

issue: #16

ghis22130 committed:
Merge pull request #21 from ghis22130/feature-Alamofire

Feature alamofire

ghis22130 committed:
feat: Mod BanchanListViewMdoel, BanchanListViewController

DiffableDataSource를 이용한 Section Update 가능하도록 구현하였습니다.
Notification을 이용하였지만 Combine으로 구현해볼 예정입니다.

issue: #13, #16

ghis22130 committed:
Merge pull request #23 from ghis22130/feature-diffable

feat: Mod BanchanListViewMdoel, BanchanListViewController

youngminshim-de committed:
refactor: BanchanListViewController

Notification을 이용하여 collectionView를 업데이트 하던 방식에서 Combine을 활용한 방식으로 변경

youngminshim-de committed:
feat: Add Toaster

토스터를 이용하여 제목을 누르면 상품의 개수가 화면에 나오도록 구현
issue:#15

youngminshim-de committed:
feat: Add Image Load

이미지를 가져와 앱에 뿌려지도록 구현

issue:#12

youngminshim-de committed:
Merge pull request #24 from ghis22130/feature-Combine

Feature combine

youngminshim-de committed:
refactor: BanchanListViewController Refactoring

BanchanListViewController 의 Datasource 에서 작업하던 cell 의 image 업데이트를 cell이 직접하도록 수정

youngminshim-de committed:
Merge pull request #25 from ghis22130/feature-Combine

refactor: BanchanListViewController Refactoring

napster-x-90 commented on iOS/SideDish/SideDish/Data/Network/DataMapping/BanchanListDTO+Mapping.swift:
```diff
@@ -0,0 +1,53 @@
+        let salePrice: String
+        let badge: [String]?
+        
+        private enum CodingKeys: String, CodingKey {
```
CodingKey를 따로 사용하신 이유가 있을까요?
없다면 카멜케이스로 Decoding할 수 있는 옵션이 있습니다. 😀

ghis22130 replied:
카멜케이스로 Deconding 하고자 하는 의도로 도입했었는데 `convertFromSnakeCase` 가 존재 하는것을 알았네요. 
적용하면 불필요한 enum을 줄일 수 있어 코드가 더욱 깔끔해질 것 같습니다! 감사합니다.😀

napster-x-90 commented on iOS/SideDish/SideDish/Data/Network/DataMapping/BanchanListDTO+Mapping.swift:
```diff
@@ -0,0 +1,53 @@
+
+import Foundation
+
+struct BanchanListDTO: Codable {
```
Codable은 Encoding과 Decoding이 함께 채택되어있는 프로토콜입니다.
해당 구조체가 Decoding만 한다면 Decodable만 채택하는게 다른사람이 봤을때
이 객체는 Decoding만 하는구나 라고 명시적으로 알 수 있습니다.

ghis22130 replied:
해당 DTO는 Decoding만 하고있으니 Decodable을 채택하도록 수정하겠습니다.
Encoding, Decoding에 대한 명시적 구분도 중요하다는 것을 배웠습니다. 🧐

napster-x-90 commented on iOS/SideDish/SideDish/Data/Network/DataMapping/BanchanListDTO+Mapping.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+}
+
+extension BanchanListDTO.BanchanListItemDTO {
```
해당 Extension은 꼭 Extension으로 빼야할까요?
BancanListItemDTO 내부에 있는게 읽기 편해보입니다.

ghis22130 replied:
DTO 내부 property들만 명시적으로 먼저 보여주는 것이 중요하다 생각했습니다!

DTO 안에 property, enum, method 가 들어가게되다 보니 번잡해 보여서 toDomain() 메소드 들은 extension 으로 빼서 모아 주는것이 깔끔하다 여겼는데, 
위에서 리뷰해주신 `Coding Key` 관련 enum을 사용하지 않게 되면 내부로 들어가는게 더 좋을것 같군요!

napster-x-90 commented on iOS/SideDish/SideDish/Data/Network/NetworkService.swift:
```diff
@@ -0,0 +1,29 @@
+        guard let url = URL(string: strUrl) else { return }
+        
+        AF.request(url, method: httpMethod).validate().responseData { data in
+            completion(data)
```
항상 API는 성공만 하지 않습니다.
Error에대한 처리를 잊지 말아주세요 🤔

ghis22130 replied:
error에 대한 처리 ... 명심하겠습니다!  👍

napster-x-90 commented on iOS/SideDish/SideDish/Domain/Entities/Banchan.swift:
```diff
@@ -0,0 +1,37 @@
+    private (set) var netPrice: String?
+    private (set) var salePrice: String
+    private (set) var badge: [String]?
+    private (set) var delivery_type: [String]
```
카멜케이스가 필요해보이네요!!
그리고 보통 배열이라면 속성의 이름은 복수형으로 해주셔야합니다

napster-x-90 commented on iOS/SideDish/SideDish/Domain/UseCases/FetchBanchanListUseCase.swift:
```diff
@@ -0,0 +1,22 @@
+import Foundation
+
+struct FetchBanchanListUseCase {
+    private static let baseURL: String = "https://h3rb9c0ugl.execute-api.ap-northeast-2.amazonaws.com/develop/baminchan/"
```
테스트하기 좋은 코드는 이러한 값을 외부에서 주입받는 방식입니다.
이렇게되면 테스트할때 baseURL을 수정할 수 없게 됩니다 😭

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanCustomCell.swift:
```diff
@@ -0,0 +1,53 @@
+    
+    private func loadImages(imgURL: String) {
+        DispatchQueue.global().async {
+            FetchImageUseCase.fetch(network: NetworkSerivce.shared,
```
URL에서 이미지를 받아오는 Util을 만드는게 어떨까요? 해당 뷰에서 직접 응답을 받아 set해주는 것보다 따로 manager를 만들어서 set해주는게 좋아보입니다. 아니면 UIImageView의 Extenstion으로 빼는 것도 방법입니다.
[UIImageView Extension](https://www.hackingwithswift.com/example-code/uikit/how-to-load-a-remote-image-url-into-uiimageview)

이건 Use Case가 아닌 Image를 로드하기위한 모듈같습니다.
FetchImageUseCase를 다른 방식으로 바꿔보세요.

[KingFisher](https://github.com/onevcat/Kingfisher) 를 참고해보세요!

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanCustomCellHeader.swift:
```diff
@@ -0,0 +1,35 @@
+
+    @IBOutlet weak var titleLabel: UILabel!
+    static let identifier = "BanchanCustomCellHeader"
+    var cellCount = 0
```
접근제한자가 없다면 위험한 코드입니다. 꼭 넣어주세요. 만약
접근제한자가 없어야하다면 다시한번 설계해보세요 꼭 외부에 set 해야하는 필요한 속성일까요?

napster-x-90 commented on iOS/SideDish/SideDish/Domain/UseCases/FetchBanchanListUseCase.swift:
```diff
@@ -0,0 +1,22 @@
+    
+    static func fetchBanchanList(network: NetworkRequest,section: String, completion: @escaping ([Banchan]?) -> Void) {
+        let url = baseURL+section
+        network.request(url: url, httpMethod: .get) { dataDummy in
```
Combine을 숙지하셨다면, ViewModel과 View와의 바인딩 뿐만아니라 서버의 Request 에 대한 응답도
처리할 수 있습니다. 해당 UserCase에 Network응답값을 받아 Decoding하는 역할도 하고있네요.
이부분은 Network가 해야하는 부분인 것 같습니다.
UseCase는 [Banchan] 이라는 값만 알고 viewModel에 주입해줘야하지 않을까요?

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanCustomCell.swift:
```diff
@@ -0,0 +1,53 @@
+    
+    static let identifer = "BanchanCustomCell"
+    
+    override func awakeFromNib() {
```
아무것도 하지않는 코드는 삭제해주시는게 좋습니다. 😌

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanCustomCellHeader.swift:
```diff
@@ -0,0 +1,35 @@
+        cellCount = count
+    }
+    
+    @objc func headerTouched(_ sender: UITapGestureRecognizer) {
```
헤더가 직접 Toast를 보여주는것보단 헤더의 주인인 ViewController에서 불리우는게 맞아 보입니다.
어떻게 생각하시나요?

youngminshim-de replied:
해당 주체에 대해서 많은 고민을 했었는데, 저는 Header 안에 Toast 내용이 포함되어 있다고 생각해서 Header 안에 넣었습니다.
하지만 Header 안에 Toast 내용을 띄우지 않으니 ViewController가 맞는것 같습니다

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanListViewController.swift:
```diff
@@ -0,0 +1,93 @@
+    
+    @IBOutlet weak var banchanCollectionView: BanchanCollectionView!
+    
+    var viewModel = BanchanListViewModel()
```
접근제한자들을 꼭 명시해주세요!

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanListViewController.swift:
```diff
@@ -0,0 +1,93 @@
+            .sink(receiveCompletion: { (result) in
+                switch result {
+                case .failure(let error):
+                    print(error)
```
에러가 발생할때의 처리가 필요합니다. 
그리고 print는 PR코드에는 없어야 하는 코드입니다. 🥸

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/ViewModel/BanchanListViewModel.swift:
```diff
@@ -0,0 +1,55 @@
+    }
+    
+    func fetchMenu() {
+        FetchBanchanListUseCase.fetchBanchanList(network: network, section: "main", completion: { banchans in
```
User Case를 Static으로 쓰지않고 주입받는 식으로 고쳐보세요.
객체들간의 결합도를 낮추려면 의존성 주입은 필수입니다.

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/ViewModel/BanchanListViewModel.swift:
```diff
@@ -0,0 +1,55 @@
+        })
+    }
+    
+    func sectionCount() -> Int {
```
이 메소드는 Section Enum에 직접 구현하는게 더 좋아보입니다. ViewModel에 필요한 메소드일까요?

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/ViewModel/BanchanListViewModel.swift:
```diff
@@ -0,0 +1,55 @@
+        case main = "모두가 좋아하는 든든한 메인요리"
+    }
+    
+    @Published var menu: Dictionary<Section, [Banchan]>
```
언제든 클래스내부의 속성은 최소한의 접근 제한자가 필요합니다.
만약 public하게 사용해야한다면 설계를 다시해볼 필요가 있습니다!! 

Section을 dictionary로 만든다면 키값을 가지고 인덱스의 값을 계산해야하는 번거로움이 있습니다.
보통은 섹션도 인덱스로 관리를 합니다.
딕셔너리타입으로 하면 어떤 불편함이 있을까요? 한번 생각해보세요.
되도록이면 2차월 배열로 섹션을 구분해보세요. 😁

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanListViewController.swift:
```diff
@@ -0,0 +1,93 @@
+// MARK: - Delegate
+extension BanchanListViewController: UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: 343, height: 130)
```
상수를 두어 값을 직접 사용하는 것보다 나아보입니다.
`let cellSize = CGSize(width: 343, height: 130)`

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanListViewController.swift:
```diff
@@ -0,0 +1,93 @@
+                    break
+                }
+            }, receiveValue: { (value) in
+                self.applySnapshot()
```
순환참조를 방지하기위해 캡쳐리스트를 사용해주세요 
value는 사용하지 않는다면 _ 로 바꿔주시는게 더 좋겠죠? 😌

napster-x-90 commented on iOS/SideDish/SideDish/Domain/Entities/Banchan.swift:
```diff
@@ -0,0 +1,37 @@
+import Foundation
+
+enum Delivery {
+    static let dawn = "새벽배송"
```
enum으로 사용하셨다면 case의 이점을 이용하는게 좋지 않을까요? 😃

ghis22130 replied:
이 enum들에 대한 처리가 아주 고민이였습니다 😥 
case의 이점 인지하고 개선해보겠습니다!

napster-x-90 commented on iOS/SideDish/SideDish/Domain/UseCases/FetchImageUseCase.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+struct FetchImageUseCase {
```
단순히 이미지만 받아오는 건 Use Case가 아닌 것 같아요. 
:) 다른 클래스에서 설명드린대로 한번 살펴보세요.

ghis22130 replied:
이 부분도 향후 구현할 Repository, Persistent Storage와 함께 깊게 고민해봐야할 것 같습니다.
저도 구현하면서 아리송 했었는데 역시 였군요..ㅎ

MVVM에 대한 이해가 더 필요한 것  같습니다 😮

napster-x-90 commented on iOS/SideDish/SideDish/Presentaion/BanchanList/View/BanchanCustomCell.swift:
```diff
@@ -0,0 +1,53 @@
+            loadImages(imgURL: banchan?.image ?? "")
+            titleLabel.text = banchan?.title
+            descriptionLabel.text = banchan?.description
+            netPriceLabel.text = "1000"
```
가격은 고정인가요?
반찬속성에 값이 set될때 고정된 값이 보일 것같습니다.

ghis22130 replied:
UI 구현이 완전하지 않아 우선 고정적으로 두었으나 개선할 예정입니다!

napster-x-90 commented on iOS/SideDish/SideDish/Domain/Entities/DetailBanchan.swift:
```diff
@@ -0,0 +1,22 @@
+
+import Foundation
+
+struct DetailBanchan {
```
왠지 느낌일지 모르지만 BanchanDetail이 더 익숙합니다.

napster-x-90 reviewed (commented):
안녕하세요! Napster라고합니다 :) 반갑습니다.
부족하지만 최선을 다해 리뷰를 해보려합니다. 

팀원을 이루어서 작성한 코드이지만 코드 컨벤션이 잘 지켜진것 같습니다. 의도하신건가요? :)
Clean Architecture 의 도입은 아주 좋습니다. 다만,
모든 건 왜 사용하는지에 대한 나름의 기준이 분명히 있어야합니다.
배웠으니까, 모두 하니까 적용해봤습니다는 타당한 이유가 되지 않습니다. 😙
나름의 기준을 세워보세요.

ViewModel의 역할은 어디까지일까요?
BancanListViewController에 데이터를 바인딩해서 보여주고있지만,
applySnapshot 메소드가 하는 역할이  ViewModel의 역할과 조금 겹쳐보입니다.
CollectionViewDataSource와 View의 바인딩은 정말 이방법 밖에 없을까요?
Combine을 사용했다면 좀 더 명확한 표현의 바인딩이 있을 것 같습니다.
코드에 Combine을 도입이 되었는데 되도록이면 Closure사용을 줄이고  Event로 바꿔보세요!

😄 잘 알고 계신 것 같지만 싱글톤의 사용을 최소한으로 줄여보세요.
NetworkService에 대한 고민이 필요해보입니다.
Swift Network Layer를 키워드로 자료를 찾아보세요.

수고하셨습니다!!
화이팅합시다!

youngminshim-de committed:
refactor: Mod BanchanListDTO+Mapping Class

반찬리스트DTO 클래스의 코딩키 삭제 & ConvertfromSnakeCase 적용 및 변수명 변경

issue: #27

youngminshim-de committed:
refactor: Add Access Modifer

클래스 내부의 속성 접근 제한자 설정

issue:#27

youngminshim-de committed:
feat: Add BadgeStackView in BanchanCustomCell

반찬 커스텀셀에 특가 배지 넣기위해 BadgeStaciView를 추가하였습니다.

issue: #12

youngminshim-de committed:
refactor: Mod SectionList Order

섹션이 순서대로 나오지 않는 버그 수정

youngminshim-de committed:
refactor: Mod PropertyName & Delete Unnecessary func

변수명 수정 및 불필요한 함수 제거

issue: #27

youngminshim-de committed:
refactor: Mod Toaster

기존에 헤더에서 불리워지던 토스터를 ViewController에서 불리워지도록 수정
issue: #27

youngminshim-de committed:
refactor: Reflect review

변수명 및 클래스명 변경

issue: #27

youngminshim-de committed:
refactor: Mod Network & FetchBanchanUseCase

컴바인을 이용하여 네트워크 리퀘스트를 받아오도록 변경하였고, FetchBanchanUseCase 또한 컴바인을 이용하도록 변경하였다.

youngminshim-de committed:
refactor: Mod getImage in Cell

KingFisher와 ImageView Extension을 이용하여 Cell의 imageView의 image Update 하는 부분 변경

youngminshim-de committed:
refactor: Mod UseCase & ViewModel

URL을 ViewModel에서 UseCase로 주입해주는 형태로 변경하였고, UseCase에서 static method를 쓰던 방식에서 ViewModel이 UseCase를 가지고 있는 형태로 변경

youngminshim-de committed:
Merge pull request #34 from ghis22130/feature-CoreData

Refacotring: FeedBack 내용 수정 및 Network Combine으로 변경
